docs: FileSystem (#42)
NativeScript/docs: content/guide/core/file-system.md

  
})

// OR

folder.clearSync((error)=>{

})
```

### Deleting a folder

To delete a folder, use the `remove` or `removeSync` method:

```ts
folder.remove().then((value: boolean)=>{

console.log(value)

}).catch((error)=>{

console.error(error.message)
})
```

### Creating, writing to and reading from a text file
<!-- Preview: https://stackblitz.com/edit/nativescript-stackblitz-templates-5yndru?file=app/main-view-model.ts -->
- To create a file, call the `getFile()` method on an instance of the `Folder` class and pass it the file name with the extension.

```ts
folder.getFile("my-text.txt")
//or 
const filePath = path.join(folder.path, 'my-text.txt')
const file = File.fromPath(filePath)
```

- To save a text to a file, use the `writeText` or `writeTextSync` method:

```ts
file
  .writeText('Some text')
  .then(result => {
    // Succeeded writing to the file.
 
  })
  .catch(error => {
    console.log(error)
  })

```
- To read data from a text file, use the `readText` or `readTextSync` method of the file instance:
```ts
   file.readText().then(res => {
      // Succeeded read from file.
      
    })
    .catch(error => {

  })
```
### Check if a file exists

> To check if a file exists, call the static [exists](#exists) method of the File class and pass it the file path string:

Reviewer: Typo.

"pass it the file path string"

Should be "pass it to the file path stored in a string"

Author: I think it should be "pass it the file path as a string"

Reviewer: "pass the method the file path as a string"
No confusion that way.

```ts
const exists = File.exists(fPath)
```

### Renaming a file

To rename a file, use the `rename` or `renameSync` method: 

```ts
file.rename("new-name.ext").then((value)=>{

}).catch(err=>{

})
```

### Saving a binary data to a file
<!-- TODO: binary data example -->

> To save binary data to a file, use the `write` or `writeSync` method:

Reviewer: "To save binary data to a file, use either the `write` or `writeSync` method:"

```ts
file
  .write(binary data)
  .then(result => {
    // Succeeded writing to the file.
 
  })
  .catch(error => {
    console.log(error)
  })

```

### Reading a binary data from a file

To read binary data, use the `read` or `readSync` method of the File instance. For example, you can save an image on the device and retrive it as follows:

```ts
async readAnImage() {

    const folder = knownFolders.documents()
    const fPath = path.join(folder.path, 'cat.png')
    const imageFile = File.fromPath(fPath)

    try {

      const imageSource = await ImageSource.fromFile('~/assets/images/download.jpeg')
      const saved: boolean = imageSource.saveToFile(fPath, "png")

      if (saved) {
        // READ BINARY DATA
        const imageBinary = await imageFile.read()
        const img = await ImageSource.fromData(imageBinary)
      }

    } catch (err) {

    }
}
```

### Deleting a file

To remove a file, use the [remove](#remove) or [removeSync](#removesync) method of the File instance:

```ts
file
  .remove()
  .then((res: boolean) => {
    // Success removing the file.
  })
  .catch(err => {
  })
 ``` 

 ### Normalizing a path

>  To normalize a path use the `normalize` or create it with the `join` method from the [path](#path-operations) object:

Reviewer: To normalize a path use the `normalize` method or create the path with the `join` method from the [path](#path-operations) object:

```ts
const testPath = '///test.txt'
let documentsFolder = knownFolders.documents()
path.normalize(documentsFolder.path + testPath)
```

## FileSystem API

### documents()

```ts
const folder: Folder = knownFolders.documents()
```

Gets the Documents folder available for the current application. This Folder is private for the application and not accessible from Users/External apps.

---

### externalDocuments()

```ts
const folder: Folder = knownFolders.externalDocuments()
```

> Gets the Documents folder available for the current application on an external storage. This Folder is private for the application and not accessible from Users/External apps. On android this requires `READ_EXTERNAL_STORAGE/WRITE_EXTERNAL_STORAGE` permissions. There is no external storage on iOS os it is the same as [documents()](#documents).

Reviewer: Gets the Documents folder available for the current application from an external storage source. This folder has private access, with availability limited to the application, meaning it is not accessible to outside users, bots or external apps. 

On Android, for such read or write access, the flags `READ_EXTERNAL_STORAGE/WRITE_EXTERNAL_STORAGE` permissions need to be set to true.  This can be done by adding the following XML code to the AndroidManifest.xml file ->
  <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />

. There is no external storage on iOS, it is the same as [documents()](#documents).

---  

### temp()

```ts
const folder: Folder = knownFolders.temp()
```
Gets the Temporary (Caches) folder available for the current application. This Folder is private for the application and not accessible from Users/External apps.

---

### currentApp()

```ts
folder: Folder = knownFolders.currentApp()
```

Gets the root folder for the current application.

---

### iOS known folders

> The following methods allow to access to iOS known folders.

Reviewer: The following methods enable access to iOS known folders.

```
Gets the NSDownloadsDirectory.

---

#### movies()
```ts
folder: Folder = knownFolders.ios.movies()
```
Gets the NSMoviesDirectory.

---

#### music()
```ts
folder: Folder = knownFolders.ios.music()
```
Gets the NSMusicDirectory.

---

#### pictures()
```ts
folder: Folder = knownFolders.ios.pictures()
```
Gets the NSPicturesDirectory.

---

#### sharedPublic()
```ts
folder: Folder = knownFolders.ios.sharedPublic()
```
Gets the NSSharedPublicDirectory.

---

### fromPath()

```ts

file: File = File.fromPath(path)

```

or 

```ts
folder: Folder = Folder.fromPath(path)
```

Gets or creates a Folder or File entity at the specified path.

---

### getFolder
```ts
folder : Folder = folder.getFolder(name)
```

Gets or creates a Folder entity with the specified `name` within a Folder.

---

### exists

```ts
folderExists: boolean = Folder.exists(path)
```

or 

```ts
file: boolean = File.exists(path)
```

Checks whether a Folder or File with the specified path already exists.

---

### isKnown

```ts
isItAKnownFolder: boolean = folder.isKnown
```
Determines whether this instance is a known folder (accessed through the `knownFolders` object).

---

> Both the File and Folder classes extend the FileSystemEntity which has the following API:

Reviewer: Both the File and Folder classes extend the FileSystemEntity, which has the following API:

```
Asynchronously gets all the top-level entities residing within a folder.

---

### getEntitiesSync()

```ts
folder.getEntitiesSync(onError?: (error: any) => any): Array<FileSystemEntity>
```

Gets all the top-level entities residing within this folder synchronously. `onError` is a optional function to be called if some error occurs.

---

### eachEntity()

```ts
folder.eachEntity(onEntity: (entity: FileSystemEntity) => boolean)
```

Enumerates all the top-level FileSystem entities within a folder. `onEntity` is a callback that receives the current entity.

---
### getFile()
```ts
folder.getFile(name: string): File
```
Gets or creates a File entity with the specified name within this Folder

---

### extension

```ts
fileExt: string = file.extension
```

Gets the extension of the file.

---

### size

```ts
fileSize: number = file.size
```

Gets the extension of the file.

---

### isLocked

```ts
isFileLocked: boolean = file.isLocked
```

Gets a boolean value indicating whether the file is currently locked, meaning a background operation associated with this file is running.

---

### readText()

```ts
text: string = await file.readText(encoding)
```

> Asynchronously reads the content of the file as a string using an optional encoding value. If you do not pass any encoding, `UTF-8` is used.

Reviewer: Asynchronously reads the content of the file as a string using an optional encoding value. If you do not pass any encoding, `UTF-8` is used by default.

---
title: FileSystem
description: Work with the device file system
---

> The FileSystem module provides high-level abstractions for file system entities such as files, folders, known folders( the entry to the device file system), paths, separators, etc.

Reviewer: The FileSystem module offers easy-to-use tools for working with files and folders in a computer's file system. It provides helpful functions for managing files, folders, paths, and separators, etc.

> The FileSystem module provides high-level abstractions for file system entities such as files, folders, known folders( the entry to the device file system), paths, separators, etc.

Reviewer 2: ```suggestion
The FileSystem module offers easy-to-use tools for working with files and folders in a computer's file system. It provides helpful functions for managing files, folders, paths, and separators, etc.
```

  })

  ```

### Deleting the contents of a Folder

To delete all the content of a Folder, use the [clear](#clear) or [clearSync](#clearsync) method of a Folder instance:

```ts
folder.clear().then((result)=>{
// successful delete
}).catch((rerror)=>{
  
})

// OR

folder.clearSync((error)=>{

})
```

### Deleting a folder

To delete a folder, use the `remove` or `removeSync` method:

```ts
folder.remove().then((value: boolean)=>{

console.log(value)

}).catch((error)=>{

console.error(error.message)
})
```

### Creating, writing to and reading from a text file
<!-- Preview: https://stackblitz.com/edit/nativescript-stackblitz-templates-5yndru?file=app/main-view-model.ts -->
- To create a file, call the `getFile()` method on an instance of the `Folder` class and pass it the file name with the extension.

```ts
folder.getFile("my-text.txt")
//or 
const filePath = path.join(folder.path, 'my-text.txt')
const file = File.fromPath(filePath)
```

- To save a text to a file, use the `writeText` or `writeTextSync` method:

```ts
file
  .writeText('Some text')
  .then(result => {
    // Succeeded writing to the file.
 
  })
  .catch(error => {
    console.log(error)
  })

```

> - To read data from a text file, use the `readText` or `readTextSync` method of the file instance:

Reviewer: How about something less dry, such as this ->

To extract data from a text file, you can use the readText or readTextSync method of the file instance. Here's an example of how to use it:

> - To read data from a text file, use the `readText` or `readTextSync` method of the file instance:

Reviewer 2: ```suggestion
- To extract data from a text file, you can use the readText or readTextSync method of the file instance. Here's an example of how to use it:
```

```ts
   file.readText().then(res => {
      // Succeeded read from file.
      
    })
    .catch(error => {

  })
```
### Check if a file exists

> To check if a file exists, call the static [exists](#exists) method of the File class and pass it the file path string:

Reviewer 2: ```suggestion
To check if a file exists, you can use [exists](#exists) and pass it the file path as a string:
```

```ts
file
  .write(binary data)
  .then(result => {
    // Succeeded writing to the file.
 
  })
  .catch(error => {
    console.log(error)
  })

```

### Reading a binary data from a file

To read binary data, use the `read` or `readSync` method of the File instance. For example, you can save an image on the device and retrive it as follows:

```ts
async readAnImage() {

    const folder = knownFolders.documents()
    const fPath = path.join(folder.path, 'cat.png')
    const imageFile = File.fromPath(fPath)

    try {

      const imageSource = await ImageSource.fromFile('~/assets/images/download.jpeg')
      const saved: boolean = imageSource.saveToFile(fPath, "png")

      if (saved) {
        // READ BINARY DATA
        const imageBinary = await imageFile.read()
        const img = await ImageSource.fromData(imageBinary)
      }

    } catch (err) {

    }
}
```

### Deleting a file

To remove a file, use the [remove](#remove) or [removeSync](#removesync) method of the File instance:

```ts
file
  .remove()
  .then((res: boolean) => {
    // Success removing the file.
  })
  .catch(err => {
  })
 ``` 

 ### Normalizing a path

>  To normalize a path use the `normalize` or create the path with the `join` method from the [path](#path-operations) object:

Reviewer: Is it worthwhile mentioning what normalizing a path means?
